enterprise/backend/src/metabase_enterprise/metabot_v3/tools/find_outliers.clj

```diff
@@ -1,8 +1,42 @@
 (ns metabase-enterprise.metabot-v3.tools.find-outliers
   (:require
+   [medley.core :as m]
+   [metabase-enterprise.metabot-v3.client :as metabot-v3.client]
    [metabase-enterprise.metabot-v3.tools.interface :as metabot-v3.tools.interface]
-   [metabase.util.malli :as mu]))
+   [metabase.api.common :as api]
+   [metabase.lib.types.isa :as lib.types.isa]
+   [metabase.query-processor :as qp]
+   [metabase.util :as u]
+   [metabase.util.malli :as mu]
+   [toucan2.core :as t2]))
 
 (mu/defmethod metabot-v3.tools.interface/*invoke-tool* :metabot.tool/find-outliers
-  [_tool-name _arguments _env]
-  {:output "Not implemented."})
+  [_tool-name {{:keys [query_id
+                       report_id
+                       metric_id
+                       result_field_id]}
+               :data-source} _env]
+  (when (or query_id report_id result_field_id)
+    (throw (ex-info "Not implemented" {})))
+  (let [{:keys [dataset_query]} (api/read-check (t2/select-one :model/Card metric_id))
+        {:keys [data status]} (qp/process-query dataset_query)]
+    (when-not (= :completed status)
+      (throw (ex-info "Unexpected error running query" {})))
+    (let [dimension-col-idx (->> data
+                                 :cols
+                                 (map-indexed vector)
+                                 (m/find-first (fn [[_i col]]
+                                                 (lib.types.isa/temporal? (u/normalize-map col))))
+                                 first)
+          value-col-idx (->> data
+                             :cols
+                             (map-indexed vector)
+                             (m/find-first (fn [[_i col]]
+                                             (lib.types.isa/numeric? (u/normalize-map col))))
+                             first)]
```

Comment:
This might be good enough for our current experiments, but it won't work in general: metrics can have no or several breakouts and the breakouts don't have to be temporal.

Reply:
But they always have a default time dimension, right?
I was thinking when you just randomly want to get outliers for a metric, then we use the default time dimension that a user sets when creating the metric (this is what we do here, right @johnswanson ?).

When instead, you want to get outliers on the results of a `query-metric` (e.g. when you did breakouts, etc.), then we would want to see outliers for that specific result (`query_id` + `result_field`). The latter still has some flaws as we [identified](https://metaboat.slack.com/archives/C07SJT1P0ET/p1734615972594109?thread_ts=1734601795.489089&cid=C07SJT1P0ET)

Reply:
I've gone and checked. The FE doesn't currently let you break out by non-temporal columns (but the BE would allow and handle that). It does allow defining metrics without any breakouts (i.e., dimensions).

> But they always have a default time dimension, right?

No, you can define a metric without any breakouts.

Reply:
Ah, I didn't realize that you could create a metric without any breakouts at all. In that case the result will always be a scalar, right? And it wouldn't make sense to ask for "outliers" in that case?

Maybe it would be reasonable for the tool to just respond with an error about the problem in either of these cases:

- if somehow someone has created a metric with a non-temporal breakout (maybe there was a time when the FE didn't enforce this or something) we say something like "outliers can only be detected when a temporal dimension is available, we can't generate outliers for this metric."

- if someone asks for outliers on a metric without a breakout at all, we could say something like "outliers can only be detected when a temporal dimension is available, we can't generate outliers for this metric."

Actually as I think about it more, maybe if:
- the LLM could request to find outliers on a query with a `query_id` + `result_field` + `temporal_field`, and
- the LLM can use `query_metric` to add a breakout to a metric,

then it may even respond appropriately in this case, asking the user which field they want to use as the temporal dimension?